### DISCUSSION TITLE
connected components area not working

### DISCUSSION THREAD

**Discussion by srini8080** [ImageMagick/ImageMagick#4207] [Help]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched [open](https://github.com/ImageMagick/ImageMagick/issues) and [closed](https://github.com/ImageMagick/ImageMagick/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported
- [X] I have verified that I am using the latest version of ImageMagick

### ImageMagick version

6.9.10-71

### Operating system

Linux

### Operating system, version and so on

Centos 7.8

### Description

I have the following scanned image. I am getting 3components of the image. how do I get just  one image out of this?

### Steps to Reproduce

convert tmp_mask.png \
	-define connected-components:verbose=true \
	-define connected-components:area-threshold=9000 \
	-define connected-components:mean-color=true \
	-connected-components 4 -auto-level \
	null: 

### Images

![photos_00090](https://user-images.githubusercontent.com/12046302/120593732-88bca880-c45d-11eb-847d-7e4844d17a57.png)
![photos_00070](https://user-images.githubusercontent.com/12046302/120593738-89edd580-c45d-11eb-8889-80d75d402a6e.png)
![photos_00020](https://user-images.githubusercontent.com/12046302/120593741-89edd580-c45d-11eb-80df-911eab5d3226.png)
![pppx-080](https://user-images.githubusercontent.com/12046302/120593744-8a866c00-c45d-11eb-8c33-e512eda1ed68.png)

**Comment by snibgo** [comment]
You show three images. Why? Which one was your input? What was your text output?

**Comment by srini8080** [comment]
> You show three images. Why? Which one was your input? What was your text output?

![pppx-080](https://user-images.githubusercontent.com/12046302/120649723-b411b880-c49a-11eb-9012-439c2627fea1.png)

This is the input

**Comment by srini8080** [comment]
Output.

![photos_00090](https://user-images.githubusercontent.com/12046302/120649856-d6a3d180-c49a-11eb-96d4-3922926dc0c8.jpg)
![photos_00020](https://user-images.githubusercontent.com/12046302/120649860-d7d4fe80-c49a-11eb-89f5-e8f63b959bf4.jpg)
![photos_00010](https://user-images.githubusercontent.com/12046302/120649862-d86d9500-c49a-11eb-9061-eadff085faaf.jpg)

**Comment by srini8080** [comment]
I was only trying to get the image part of it. 

Thanks

**Comment by snibgo** [comment]
You showed this command:
```
convert tmp_mask.png
-define connected-components:verbose=true
-define connected-components:area-threshold=9000
-define connected-components:mean-color=true
-connected-components 4 -auto-level
null:
```
That command makes no images. I don't understand how you get three output images.

It does make text, from "verbose=true". If you don't want the text, use "verbose=false".

If you want an output image, then change "null:" to a filename such as "output.png". When I do that, there are clear differences between input and output, for example small black punctuation marks are removed.

**Comment by srini8080** [comment]
I want to use the connected components to extract only the image. How can I do that? I am trying to extract images out of 100 scanned pages with images and text

**Comment by snibgo** [comment]
Some connected components are larger than the letters, so what you want, "to use the connected components to extract only the image", isn't possible.

If you look at the text output, or the image, you will see that very few black marks have been removed. Those are the smallest ones. To remove more, you need to increase the threshold.

Using a threshold of 40000 removes all the individual letters, but not those that have joined together. A threshold of 60000 does remove all the letters.

**Comment by srini8080** [comment]
Is there a script I can use that would make this threshold automagically that will work for any scanned paper and only extract images?

Thanks

**Comment by snibgo** [comment]
As I say, there is no threshold that will remove all your text without also removing parts of the image.

**Comment by srini8080** [comment]
Is there something that can be done to recognize text and remove only text? Any OCR technique?

**Comment by snibgo** [comment]
A script could examine the text output of the components, looking at the gray(0) entries, ie the black components. It could accumulate the minimum and maximum of the limits of the bounding boxes. It would need to ignore the horizontal black line near the top. This would give the bounding box of the image, and from that you could do a crop.

**Comment by srini8080** [comment]
I am not very conversant with the commands. Any starter is highly appreciated.

Thanks,

**Comment by fmw42** [comment]
I do not have a full solution, only a suggestion. Average you image down to one column, then expand to say 500 columns again and threshold. You will see a pattern of stripes where there is text and a large dark area where you have an image.

Example from your image above.

`magick image.png -scale 1x! -scale 500x! -auto-threshold otsu y.png`

Here is a screen snap

![Screen Shot 2021-06-03 at 4 26 08 PM](https://user-images.githubusercontent.com/13458218/120724328-71df7b80-c488-11eb-9e30-e8adddf280c3.png)

**Comment by srini8080** [comment]
I am using linux and imagemagick 6.9.10-71

It says -auto-threshold option not found

**Comment by fmw42** [comment]
-auto-threshold is only available in ImageMagick 7. You will need to upgrade to use that feature.

**Comment by srini8080** [comment]
Thank you Alan and Fred. Making the strip worked.

I figured that Imagemagick is better than any AI/ML tool out there ;-). Thanks to Math

**Comment by srini8080** [comment]
> Some connected components are larger than the letters, so what you want, "to use the connected components to extract only the image", isn't possible.
> 
> If you look at the text output, or the image, you will see that very few black marks have been removed. Those are the smallest ones. To remove more, you need to increase the threshold.
> 
> Using a threshold of 40000 removes all the individual letters, but not those that have joined together. A threshold of 60000 does remove all the letters.

Is there a way to use the text erasing method you posted years back, to erase the text and fill the holes with white. That way there will be just the image on white background?
